pkg/agent/controller/ipseccertificate/ipsec_certificate_controller.go

```diff
@@ -0,0 +1,520 @@
+	defer klog.Infof("Shutting down %s", controllerName)
+
+	// load the existing certificates from files before starting the rotation goroutine.
+	if err := c.syncConfigurations(); err != nil {
+		klog.ErrorS(err, "Sync configurations error")
+	}
+
+	go wait.Until(c.worker, time.Second, stopCh)
+
+	go wait.Until(func() {
+		if err := c.watchFileChanges(stopCh); err != nil {
+			klog.ErrorS(err, "Failed to watch cert and key file, will retry later")
+		}
+	}, time.Minute, stopCh)
+
+	go wait.Until(func() {
+		if err := c.rotateCertificates(stopCh); err != nil {
+			klog.ErrorS(err, "Failed to rotate certificates")
+		}
+	}, time.Minute, stopCh)
+	<-stopCh
+}
+
+func (c *Controller) rotateCertificates(endCh <-chan struct{}) error {
+	for {
+		currentCert := c.certificateKeyPair.Load().(*certificateKeyPair)
+		deadline := currentCert.nextRotationDeadline()
+		if sleepInterval := time.Until(deadline); sleepInterval > 0 {
+			timer := time.NewTimer(sleepInterval)
+			defer timer.Stop()
+			select {
+			case <-endCh:
+				return nil
+			case <-c.renewEventChan:
+				// Received update event. Check the next rotation deadline again.
+				timer.Stop()
+				continue
+			case <-timer.C:
+			}
+		}
+		klog.InfoS("Starting rotating certificates")
+		csrName := fmt.Sprintf("%s-ipsec", c.nodeName)
+		err := c.kubeClient.CertificatesV1().CertificateSigningRequests().Delete(context.TODO(), csrName, metav1.DeleteOptions{})
+		if err != nil && !errors.IsNotFound(err) {
+			return fmt.Errorf("failed to delete previous CSR: %v", csrName)
+		}
+		csr, err := newCSR(csrName, c.nodeName, currentCert.privateKey)
+		if err != nil {
+			return err
+		}
+		csr, err = c.kubeClient.CertificatesV1().CertificateSigningRequests().Create(context.TODO(), csr, metav1.CreateOptions{})
+		if err != nil {
+			return err
+		}
+		ctx, cancel := context.WithTimeout(context.Background(), certificateWaitTimeout)
+		defer cancel()
+		issuedCertificate, err := csrutil.WaitForCertificate(ctx, c.kubeClient, csr.Name, csr.UID)
+		if err != nil {
+			return err
+		}
+		err = c.updateCertificates(issuedCertificate)
+		if err != nil {
+			klog.ErrorS(err, "Failed to update new certificates")
+			return err
+		}
+		klog.InfoS("Created new certificates for IPSec")
+	}
+}
+
+func (c *Controller) updateCertificates(data []byte) error {
+	cert, err := certutil.ParseCertsPEM(data)
+	if err != nil {
+		return fmt.Errorf("failed to parse certificate: %w", err)
+	}
+	if err := certutil.WriteCert(c.certPath, data); err != nil {
+		return err
+	}
+	existing := c.certificateKeyPair.Load().(*certificateKeyPair)
+	newCertPair := &certificateKeyPair{
+		rawCert:       data,
+		certificate:   cert,
+		rawKey:        existing.rawKey,
+		privateKey:    existing.privateKey,
+		rawCACert:     existing.rawCACert,
+		caCertificate: existing.caCertificate,
+	}
+	c.certificateKeyPair.Store(newCertPair)
+	c.queue.Add(workItemKey)
+	return nil
+}
+
+func (c *Controller) watchFileChanges(stopCh <-chan struct{}) error {
```

Comment:
why it needs to watch file change?

Reply:
originally i think it should be helpful for testing certificate rotation by deleting certificate files. I have removed the file watcher as it is not required.